Add library management system file for Assignment 4

rohitjoshixyz opened the pull request.

Commit:
Add library management system file for Assignment 4

the-spectator commented on 04_library_management_system.rb:
```diff
@@ -0,0 +1,125 @@
+  end
+
+  def issue_to_member(member)
+    if (@issued_to.nil?() || @status == "Available")
```
```suggestion
    if (@issued_to.nil? || @status == "Available")
```

the-spectator commented on 04_library_management_system.rb:
```diff
@@ -0,0 +1,125 @@
+
+  def issue_to_member(member)
+    if (@issued_to.nil?() || @status == "Available")
+      if (member.status == "No book taken")
```
@rohitjoshixyz Implement methods to check status. `member.book_taken?` or `member.book_available?` This makes our code future-proof as if you want to change status string in future we have to change it only at one place only.

the-spectator commented on 04_library_management_system.rb:
```diff
@@ -0,0 +1,125 @@
+        member.status = "Book taken"
+        @issued_to = member
+        @status = "Not Available"
+        puts "#{@name} issued to #{member.name}"
```
Avoid puts in methods that are dealing with business logic. Set the state in the method and make a different presentation method to print the status.

rohitjoshixyz replied:
I added `@message` to my `Book` class and then set it to the respective status_message.
However, since I need to print only one line, instead of defining a new method as follows: 
```
def print_message
  puts @message
end
```
I am returning `@message` from each method and then using `puts` right before the call in my `switch case` to print it.
Is that alright?

the-spectator replied:
Seems Good.

the-spectator commented on 04_library_management_system.rb:
```diff
@@ -0,0 +1,125 @@
+      @status = "Available"
+      @issued_to.status = "No book taken"
+      puts "#{@name} returned by #{member.name}"
+    end
```
Same as above.

the-spectator commented on 04_library_management_system.rb:
```diff
@@ -0,0 +1,125 @@
+class Library
+  attr_accessor :member_list, :book_list
+
+  def initialize()
```
When we are not accepting the parameters, avoid `()`

Commit:
Remove puts from business logic methods by returning  @message from each
method
Add book_available? and book_taken? methods
Remove brackets from methods with no arguments

the-spectator commented on 04_library_management_system.rb:
```diff
@@ -5,49 +5,67 @@ def initialize(name)
+    else
+      return true
+    end
+  end
```
Explicit return is not necessary.

the-spectator commented on 04_library_management_system.rb:
```diff
@@ -5,49 +5,67 @@ def initialize(name)
-    if (@issued_to.nil?() || @status == "Available")
-      if (member.status == "No book taken")
+    if (@issued_to.nil? || book_available?)
+      if (!member.book_taken?)
```
if we are using `if !condition` then prefer `unless condition`

the-spectator commented on 04_library_management_system.rb:
```diff
@@ -5,49 +5,67 @@ def initialize(name)
+      return true
+    else
+      return false
     end
```
Here also we can avoid explicit return

the-spectator commented on 04_library_management_system.rb:
```diff
@@ -77,7 +95,7 @@ def show_library_status
     @book_list.each_with_index do |book, i|
-      issue_status = (book.issued_to.nil?() || book.status == "Available" ? book.status : book.issued_to.name)
+      issue_status = (book.issued_to.nil? || book.status == "Available" ? book.status : book.issued_to.name)
       puts (i + 1).to_s + ". " + book.name + "      " + issue_status
```
Prefer String interpolation 
```suggestion
      puts "#{i + 1} . #{book.name}  #{issue_status}"
```

the-spectator commented on 04_library_management_system.rb:
```diff
@@ -112,14 +130,14 @@ def show_library_status
+    puts vishwas_library.book_list[book_number].issue_to_member(vishwas_library.member_list[member_number])
   when "3"
     puts "Enter member number who wants to return: "
     vishwas_library.print_members()
```
If the method doesn't accept parameters then avoid them.

rohitjoshixyz commented on 04_library_management_system.rb:
```diff
@@ -0,0 +1,143 @@
+    else
+      @message = "Sorry, #{@name} is not available. Please try another book"
+    end
+    return @message
```
Should I remove this explicit `return` as well?

the-spectator replied:
Yes

rohitjoshixyz replied:
Can I remove the `return @message` entirely since `@message` is the last assigned statement in each method or it is better to keep it explicit

the-spectator replied:
We can remove it.

Commit:
Use string interpolation
Remove brackets for methods with no parameters
Remove explicit return statements

the-spectator commented on 04_library_management_system.rb:
```diff
@@ -0,0 +1,141 @@
+  end
+
+  def book_taken?
+    if @status == "No book taken"
+      false
+    else
+      true
+    end
```
Nit pick:
```suggestion
    @status != "No book taken"
```
